### ISSUE TITLE
[API DESIGN REVIEW] and feature request: Fractional max pooling

### ISSUE THREAD

**Issue by ozabluda** [keras-team/keras#8439]
I'd love to have fractional max pooling ([Benjamin Graham, Fractional Max-Pooling (2014)](https://arxiv.org/abs/1412.6071)) in Keras. [Tensorflow has it](https://www.tensorflow.org/api_docs/python/tf/nn/fractional_max_pool).

As for API, not sure. One option is to adopt TF API. Another option is to simply allow non-integer `pool_size` and `stride` in existing Keras `[Max|Avg]Pooling[1-3]D` layers - my preference, but, I think, contrary to Keras layer API design style.

Maybe which API is supported depends on what can be done in other backends. I didn't find anything in CNTK. [Theano seems to have 3rd party implementation](https://github.com/diogo149/theano_fractional_max_pooling). For the reference, [Lasagne  3rd party implementation](https://github.com/Lasagne/Lasagne/issues/167) as well.

**Comment by fchollet** [comment]
> Another option is to simply allow non-integer pool_size and stride in existing Keras [Max|Avg]Pooling[1-3]D layers - my preference, but, I think, contrary to Keras layer API design style.

Why do you think that would be contrary to the API design style?

Is there conceptual continuity between integer and float values for the `pool_size` argument? E.g. if 2 shrinks the target axis by 2, does 1.5 shrink the target axis by 1.5?

**Comment by ozabluda** [comment]
>Why do you think that would be contrary to the API design style?

I am just learning Keras, I could easily be wrong. I got this impression because Keras has separate layers for everything, instead of single layers parametrized. For example `Conv[1|2|3]D` (and Pool and lots of other stuff), could be getting its type from the shape of the input and parameters such as `kernel_size` (just like there is no `Flatten2D` and such). Similarly Pooling could be parametrized by `max`, `avg`, `stochastic_max` (another thing I'd love to have), etc. `GlobalPooling` could be `pool_size=global` or something. Etc.

>Is there conceptual continuity between integer and float values for the pool_size argument? E.g. if 2 shrinks the target axis by 2, does 1.5 shrink the target axis by 1.5?

Yes. An even more interesting example is `pool_size=1/sqrt(2)` in 2D. It'll reduce the number of pixels by 2 in a good way (something which otherwise we have to use inception blocks to achieve).

**Comment by fchollet** [comment]
The guiding principle is user experience. There's also a bit of historicity.

There's a balance to be struck between having everything be a function argument, and having different functions for different things. We use argument for different tweaks on a shared use case, and we use different functions/classes for separate use cases. If you work with 2D data, that's one use case. 3D data is another one (different users, different data, different problem). Cleanly separating them is preferable. If there's a parameter that wouldn't work as an architecture hyperparameter when doing architecture search on a given problem, then it is better not to have it in layer signatures. E.g. `activation` or `units` is a common hyperparameter, but `ndim` would not make sense.

The pooling function type could indeed be parametrized, and in practice that's how it's implemented internally (`_Pooling2D` layer, etc). We don't do it because in practice pooling type is not an architecture hyperparameter you ever see when people do architecture search (unlike activations, for instance, which can often be parameterized across `relu/elu/etc`). It's always hard-coded.

Fractional vs. integer pooling would fit within "tweak on a same use that you might want to treat as a model hyperparameter".

**Comment by fchollet** [comment]
There's also conceptual containerization and a desire to keep the code base simple. If things like `ndim` where parametric, then functions like `call` and `compute_output_shape` would be a series of `if` blocks, and at that point it's simpler to have different classes.

This is also e.g. why we have separate layers for `LSTM` and `GRU` and not a `RNN(type='lstm')` layer.

**Comment by ozabluda** [comment]
>The guiding principle is user experience. 

Sure, the only question is which user experience is better. I work with 1D, 2D, 3D data. Topology-wise, my 3D networks start as direct analogs of 2D ones, with just the number and size of layers and filters changed. I don't see why my code should be changed at all.

Another example - if it's all the same (sanely parametrized) layer, I don't have to study multiple layers separately (and would enforce discipline on API design). For example, currently I don't why Conv1D docs are different than Conv2D docs. Until I study it, I won't know if why `data_format` is missing from Conv1D docs, and one-sided padding is missing from Conv2D docs. Or what is the reason several layers have only 2D version (I am guessing this is historic due to backend API).

>pooling type is not an architecture hyperparameter you ever see when people do architecture search

that's because currently it's all extremes either avg (T=inf) or max (T=0). Had it been properly parametrized in the frameworks some people would use (stochastic) softmax pooling with temperature (1 or 2 hyperparameters, of which avg and max are just extremes of i.e unlikely to be optimal.

>If things like `ndim` where parametric, then methods like call and `compute_output_shape` would be a series of if blocks

maybe the dynamic dispatch could all be automated. Just assign functions to variables in the preamble (or a whole "vtable"), and unconditionally call them later later, or whatever the most Pythonic approach to generics is. This could probably be easily demonstrated on a `Conv[1-3]D` class level (maybe as a hack in user code).

**Comment by ozabluda** [comment]
Fractional Max-Pooling is in statistical tie for the record on CIFAR-10 since 2015:
http://rodrigob.github.io/are_we_there_yet/build/classification_datasets_results.html#43494641522d3130
